### PR TITLE
[3.9] bpo-34480: fix bug where match variable is used prior to being defined (GH-17643)

### PR DESCRIPTION
**Pull request by mareksuscak**
This is technically not a backport of the [referenced PR](https://github.com/python/cpython/pull/17643) because the original PR was [superseded by another one](https://github.com/python/cpython/pull/8562) and as such was not accepted. I was asked to base this change on the original one because the one that superseded it and was implemented in Python 3.10 was larger in scope and technically considered a breaking change.

This PR fixes [bpo-34480](https://bugs.python.org/issue34480). I'm adding two unit tests, one of which is based on a test case included in one of the comments and I want to give full credit for writing it to the original author Karthikeyan Singaravelan (xtreak).

https://bugs.python.org/issue34480

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@mareksuscak

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by mareksuscak** [comment]
CLA is now signed. Can take up to 1 business day.

File: Lib/test/test_htmlparser.py
Comment (ezio-melotti) [inline Lib/test/test_htmlparser.py]:
```suggestion
        # bpo-34480: check that subclasses that define an
        # error method that raises an exception work
        class InvalidMarkupException(Exception):
            pass
        class MyHTMLParser(html.parser.HTMLParser):
            def error(self, message):
                raise InvalidMarkupException(message)
        parser = MyHTMLParser()
```

File: Lib/test/test_htmlparser.py
Comment (ezio-melotti) [inline Lib/test/test_htmlparser.py]:
```suggestion
        # bpo-34480: check that subclasses that define an
        # error method that doesn't raise an exception work
        class MyHTMLParser(html.parser.HTMLParser):
            def error(self, message):
                pass
        parser = MyHTMLParser()
        self.assertEqual(parser.feed('<![invalid>'), None)
```

**Review by ezio-melotti — CHANGES_REQUESTED** [review]
I added a couple of comments that point to the original issue and removed some exrta whitespace.

**Comment by mareksuscak** [comment]
@ezio-melotti thank you! I have to leave a comment per bot suggestion, to unblock this PR.

**Comment by mareksuscak** [comment]
I have made the requested changes; please review again

**Comment by ezio-melotti** [comment]
> Do you want me to add some notes to the NEWS.d/ folder? If so, can you please point me to the right file?

Yes, a NEWS entry should be added.  See https://devguide.python.org/committing/#updating-news-and-what-s-new-in-python

**Comment by mareksuscak** [comment]
Alright @ezio-melotti, I've just added a new entry in the NEWS.d folder. Please review and let me know if it looks good to you or if there's anything else that I need to do to get this bugfix merged in.

**Comment by mareksuscak** [comment]
I have made the requested changes; please review again

File: Misc/NEWS.d/next/Library/2022-04-09-11-28-49.bpo-34480.Pw-GJ6.rst
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2022-04-09-11-28-49.bpo-34480.Pw-GJ6.rst]: Please rewrite the NEWS entry. It should be formulated in terms understandable for common user. Reference to variable is a reference to internal details about which a common user know nothing.

The issue was that in some circumstances you get an UnboundVariableError. The fix is that you no longer get an UnboundVariableError. Describe these circumstances and behavior changes.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

**Comment by ambv** [comment]
We're releasing the last bugfix release of 3.9 today and this PR's been stale for a while on the NEWS entry alone. I'll commandeer the changes to the NEWS entry to land this.

**Comment by ezio-melotti** [comment]
Thanks @mareksuscak for the PR and @ambv for fixing the news entry and merging it -- I missed that this PR was waiting for my review!

**Pull request merged**
